File: Java/프로그래머스/3/42628. 이중우선순위큐/이중우선순위큐.java
import java.util.*;

class Solution {
    public int[] solution(String[] operations) {
        int[] answer = new int[2];
        PriorityQueue<Integer> minQ = new PriorityQueue<>();
        PriorityQueue<Integer> maxQ = new PriorityQueue<>(Collections.reverseOrder());

        for (String oper : operations) {
            String[] split = oper.split(" ");
            if (split[0].equals("I")) {
                minQ.offer(Integer.parseInt(split[1]));
                maxQ.offer(Integer.parseInt(split[1]));
            } else {
                if (Integer.parseInt(split[1]) == 1) {
                    Integer maxN = maxQ.poll();
                    if (maxN != null) {
                        minQ.remove(maxN);
                    }
                } else {
                    Integer minN = minQ.poll();
                    if (minN != null) {
                        maxQ.remove(minN);
                    }
                }
            }
        }

        answer[0] = maxQ.peek() == null ? 0 : maxQ.poll();
        answer[1] = minQ.peek() == null ? 0 : minQ.poll();

        return answer;
    }
}

File: Python/프로그래머스/3/42628. 이중우선순위큐/이중우선순위큐.py
import heapq


def solution(operations):
    min_q = []
    max_q = []

    for oper in operations:
        order, val = oper.split()
        val = int(val)

        if order == "I":
            heapq.heappush(min_q, val)
            heapq.heappush(max_q, -val)
        elif order == "D":
            if val == 1:
                if max_q:
                    max_val = -heapq.heappop(max_q)
                    min_q.remove(max_val)
            else:
                if min_q:
                    min_val = heapq.heappop(min_q)
                    max_q.remove(-min_val)

    if not min_q:
        return [0, 0]

    return [-heapq.heappop(max_q), heapq.heappop(min_q)]